File: IDATT2106 Systemutvikling 2 med smidig prosjekt/backend/src/main/java/com/ntnu/gidd/security/token/JwtTokenFactory.java
package com.ntnu.gidd.security.token;


import com.ntnu.gidd.security.UserDetailsImpl;
import com.ntnu.gidd.security.config.JWTConfig;
import io.jsonwebtoken.Claims;
import io.jsonwebtoken.Jws;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.SignatureAlgorithm;
import lombok.AllArgsConstructor;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.stereotype.Component;

import java.util.Arrays;
import java.util.Date;
import java.util.UUID;

/**
 * Factory class to create jwt tokens.
 */
@Component
@AllArgsConstructor
public class JwtTokenFactory implements TokenFactory {

    private JWTConfig jwtConfig;

    @Override
    public JwtAccessToken createAccessToken(UserDetailsImpl userDetails) {
        String token = Jwts.builder()
                .setSubject(userDetails.getUsername())
                .setIssuedAt(new Date(System.currentTimeMillis()))
                .setExpiration(new Date(System.currentTimeMillis() + jwtConfig.getExpiration()))
                .claim("uuid", userDetails.getId())
                .signWith(SignatureAlgorithm.HS512, jwtConfig.getSecret())
                .compact();

        return new JwtAccessToken(token);
    }

    @Override
    public JwtToken createRefreshToken(UserDetailsImpl userDetails) {
        String token = buildRefreshToken(userDetails);

        Jws<Claims> claims = Jwts.parser()
                .setSigningKey(jwtConfig.getSecret())
                .parseClaimsJws(token);

        return new JwtRefreshToken(token, claims);
    }

    private String buildRefreshToken(UserDetailsImpl userDetails) {
        Claims claims = Jwts.claims()
                .setSubject(userDetails.getUsername());
        claims.put("scopes", Arrays.asList(Scopes.REFRESH_TOKEN.scope()));

        return Jwts.builder()
                .setClaims(claims)
                .setId(UUID.randomUUID()
                               .toString())
                .setIssuedAt(new Date(System.currentTimeMillis()))
                .setExpiration(new Date(System.currentTimeMillis() + jwtConfig.getRefreshExpiration()))
                .claim("uuid", userDetails.getId())
                .signWith(SignatureAlgorithm.HS512, jwtConfig.getSecret())
                .compact();
    }
}

File: IDATT2105 Full-stack applikasjonsutvikling/Rombestilling/backend/src/main/kotlin/ntnu/idatt2105/security/token/JwtTokenFactory.kt
package ntnu.idatt2105.security.token

import io.jsonwebtoken.Claims
import io.jsonwebtoken.Jws
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.SignatureAlgorithm
import ntnu.idatt2105.core.util.SecurityConstants.AUTHORITIES_KEY
import ntnu.idatt2105.security.config.JWTConfig
import ntnu.idatt2105.user.service.UserDetailsImpl
import org.springframework.stereotype.Component
import java.util.*
import java.util.stream.Collectors

@Component
data class JwtTokenFactory(val jwtConfig: JWTConfig) : TokenFactory {

    @Suppress("DEPRECATION")
    override fun createAccessToken(userDetails: UserDetailsImpl): JwtAccessToken {
        val authorities = userDetails.authorities.stream()
            .map { it?.authority }
            .collect(Collectors.joining(","))

        val token: String = Jwts.builder()
                .setSubject(userDetails.username)
                .setIssuedAt(Date(System.currentTimeMillis()))
                .setExpiration(Date(System.currentTimeMillis() + jwtConfig.expiration))
                .claim("uuid", userDetails.getId())
                .claim(AUTHORITIES_KEY, authorities)
                .signWith(SignatureAlgorithm.HS512, jwtConfig.secret)
                .compact()

        return JwtAccessToken(token)
    }

    override fun createRefreshToken(userDetails: UserDetailsImpl): JwtToken {
        val token = userDetails?.let { buildRefreshToken(it) }
        val claims: Jws<Claims> = Jwts.parserBuilder()
                .setSigningKey(jwtConfig.secret).build()
                .parseClaimsJws(token)
        return token?.let { JwtRefreshToken(it, claims) }
    }
    @Suppress("DEPRECATION")
    private fun buildRefreshToken(userDetails: UserDetailsImpl): String {
        val claims: Claims = Jwts.claims()
                .setSubject(userDetails.getUsername())
        claims.put("scopes", Arrays.asList(Scopes.REFRESH_TOKEN.scope()))

        return Jwts.builder()
                .setClaims(claims)
                .setId(UUID.randomUUID()
                        .toString())
                .setIssuedAt(Date(System.currentTimeMillis()))
                .setExpiration(Date(System.currentTimeMillis() + jwtConfig.refreshExpiration))
                .claim("uuid", userDetails.getId())
                .signWith(SignatureAlgorithm.HS512, jwtConfig.secret)
                .compact()
    }
}
